### ISSUE TITLE
UART: Add protocol

### ISSUE THREAD

**Issue by wallento** [TUM-LIS/glip#4]
One of the great things of glip is the availability of a reset signal. This one is currently lacking with UART at the moment. As there is no out-of-band control signals, we have to do it in-band, meaning we need to put up a simple flow control protocol. This has the advantage that we can also support UARTs without RST/CTS flow control.
## Why not in-band control words?

A simple option to add in-band control words is to define one word as special. It is then necessary to filter for this word and check the next word coming in. If it is the same again, it was a data item of that value, a control otherwise. This reduces the data rate a bit as for this special word data items are duplicated. Hence, it should probably not be 0x0, 0xff or any low numbers due to their high occurrence rate in common data streams. 0xfe may be a good candidate.

While this works good in hardware it needs some attention in software as we need to scan data arrays for 0xfe and handle it properly. But especially for UART this is not critical in timing.

But: We have a problem with backpressure from the logic. If the logic deadlocks for any reason that might be especially a case for a reset. But how do you get it through now? First, we should flush the Linux output buffer to get the reset to the top of it. But then the logic has to get it, but there is no way to accept a reset but not other words. If a word gets through it needs to be forwarded before getting the next one. Hence, the reset will never get through. Dropping words (maybe with a timeout) are also not good options.
## Protocol, credit-based

Similar to the JTAG backend we need a simple backend that does flow control by sending the number of words the input FIFO can accept to the host (credit). The host is only allowed to send that many words to the logic. It can then send in a control word that is handled by the backend logic directly at the input.
### Credit word (logic->host)

The logic sends a credit word to the host allowing it to send that many words to the logic. The problem is again how to distinguish that from the actual data stream. Here we use the special word 0xfe now. The logic sends a credit as:

```
+------+-------------------+-------------+
| 0xfe | {credit[14:8], 1} | credit[7:0] |
+------+-------------------+-------------+
```

This allows to give credit up to 32767 words, what should be sufficient. If the backend logic detects a 0xfe in the data stream it duplicates it as (byte 0 in the second word is 0 then). The host software can extract it then.

An open question is when to send the credit word. Options are: time-based or threshold-based.
### Data packets and control packets (host->logic)

The host can now send arbitrary large packets by sending an indicator packet before it, that gives the length. The logic now knows that there are that many data items to forward. Up to 254 data items can be send after it.

If the first word is 0xff, the next word after is a control word. Currently only reset should be supported.

Finally, if the first word is 0x0, it should be ignored.
## Throughput considerations

The host->logic throughput depends on the average item size. If the data packets are all full there will 1 out of 255 packets overhead, which is 0.4 percent loss. With smaller packets it can grow up to 50 percent of course. Hence it is desirable to aggregate multiple words in a buffer before sending them to the logic. This may be done with a timer or termios magic. Waiting for a few words for a few milliseconds is not that much of a loss compared to the slow UART speed.

The logic->host overhead depends on the first on the occurrence of 0xfe in the data stream. If it is equally distributed, then there is 1 out of 257 packets overhead. The credit messages are preferably rare, but need further consideration.

**Comment by wallento** [comment]
Actually, it is not sufficient to send the credit in the data stream, as it requires that there is a data stream. Hence, we need to run a similar protocol in the logic->host direction.

**Comment by wallento** [comment]
Solved by 620330cb5f7e619af3e254021c72b82ae6dab2db
